File: csharp/Mcp.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;

namespace Hermes
{
    public class McpPlugin
    {
        public string Name { get; set; }
        public Process Process { get; set; }
        public StreamWriter StandardInput { get; set; }
        public StreamReader StandardOutput { get; set; }
    }

    public class McpPluginManager
    {
        private readonly Dictionary<string, McpPlugin> _plugins = new Dictionary<string, McpPlugin>();
        private readonly object _lock = new object();

        public void Load(string name, string command, string args)
        {
            lock (_lock)
            {
                var startInfo = new ProcessStartInfo
                {
                    FileName = command,
                    Arguments = args,
                    RedirectStandardInput = true,
                    RedirectStandardOutput = true,
                    UseShellExecute = false,
                    CreateNoWindow = true
                };

                var process = Process.Start(startInfo);
                if (process == null) throw new Exception("Failed to start process");

                var plugin = new McpPlugin
                {
                    Name = name,
                    Process = process,
                    StandardInput = process.StandardInput,
                    StandardOutput = process.StandardOutput
                };

                _plugins[name] = plugin;
                Console.WriteLine($"[hermes:mcp] Loaded plugin: {name}");
            }
        }

        public void Unload(string name)
        {
            lock (_lock)
            {
                if (_plugins.TryGetValue(name, out var plugin))
                {
                    plugin.Process.Kill();
                    _plugins.Remove(name);
                    Console.WriteLine($"[hermes:mcp] Unloaded plugin: {name}");
                }
            }
        }
    }
}

File: typescript/mcp.ts
import { spawn, ChildProcessWithoutNullStreams } from 'child_process';

export class McpPlugin {
    constructor(
        public name: string,
        public process: ChildProcessWithoutNullStreams
    ) {}
}

export class McpPluginManager {
    private plugins: Map<string, McpPlugin> = new Map();

    public load(name: string, command: string, args: string[]): void {
        const proc = spawn(command, args);

        const plugin = new McpPlugin(name, proc);
        this.plugins.set(name, plugin);
        console.log(`[hermes:mcp] Loaded plugin: ${name}`);
    }

    public unload(name: string): void {
        const plugin = this.plugins.get(name);
        if (plugin) {
            plugin.process.kill('SIGKILL');
            this.plugins.delete(name);
            console.log(`[hermes:mcp] Unloaded plugin: ${name}`);
        }
    }
}
